lightning-background-processor/src/lib.rs

```diff
@@ -158,13 +160,43 @@ impl BackgroundProcessor {
 				}
 			}
 		});
-		Self { stop_thread: stop_thread_clone, thread_handle: handle }
+		Self { stop_thread: stop_thread_clone, thread_handle: Some(handle) }
+	}
+
+	/// Join `BackgroundProcessor`'s thread, returning any error that occurred while persisting
+	/// [`ChannelManager`].
+	///
+	/// [`ChannelManager`]: lightning::ln::channelmanager::ChannelManager
+	pub fn join(mut self) -> Result<(), std::io::Error> {
+		assert!(self.thread_handle.is_some());
+		self.join_thread()
+	}
+
+	/// Stop `BackgroundProcessor`'s thread, returning any error that occurred while persisting
+	/// [`ChannelManager`].
+	///
+	/// [`ChannelManager`]: lightning::ln::channelmanager::ChannelManager
+	pub fn stop(mut self) -> Result<(), std::io::Error> {
+		assert!(self.thread_handle.is_some());
+		self.stop_and_join_thread()
 	}
 
-	/// Stop `BackgroundProcessor`'s thread.
-	pub fn stop(self) -> Result<(), std::io::Error> {
+	fn stop_and_join_thread(&mut self) -> Result<(), std::io::Error> {
 		self.stop_thread.store(true, Ordering::Release);
-		self.thread_handle.join().unwrap()
+		self.join_thread()
+	}
+
+	fn join_thread(&mut self) -> Result<(), std::io::Error> {
+		match self.thread_handle.take() {
+			Some(handle) => handle.join().unwrap(),
```

Comment:
It seems that if the background thread panics, `join` will return an error that will cause the `unwrap` to panic on this line. Might it'd be preferable to either return the panic as an error or at least document that this is the expected behavior?

Reply:
Added docs about panicking to `join` and `stop`.

Reply:
It does feel a bit strange that we have a function that returns a `Result`, but instead of ever returning an `Err` it always panics.

Reply:
Hmm, so the behavior of `std`'s `join()` is to return an error if the thread panics. Might it make more sense for our `join` method to mirror that behavior?

Reply:
> It does feel a bit strange that we have a function that returns a `Result`, but instead of ever returning an `Err` it always panics.

It will return an `Err` if the persister returns an `Err`. The `unwrap` is to remove the error handling `JoinHandle` provides for panics, not for the `Result` returned by the thread's closure.

> Hmm, so the behavior of `std`'s `join()` is to return an error if the thread panics. Might it make more sense for our `join` method to mirror that behavior?

My intention was to mirror the same behavior as `stop`, which would already panic in this case. The reasoning for panicking was because the panic would likely have originated from user code. So users should either write code to not panic in the case of event handling (some discussion [here](https://github.com/rust-bitcoin/rust-lightning/pull/920#discussion_r631399789)) or to return an error in case of persisting. If the panic is coming from RL, my assumption was we'd want to propagate it.

I'm not necessarily opposed to making it return the wrapped result, but I think we'd want `stop` to have the same behavior. There may be a good argument based on how a user configures Rust to handle panics. I vaguely remember @TheBlueMatt discussing it with us but I'm a bit foggy on the details.